### ISSUE TITLE
Troubles updating to 3.0.2

### ISSUE THREAD

**Issue by codyjames** [putyourlightson/craft-snaptcha#15]
### Describe the bug
I'm trying to update to the newest version of Snaptcha and am having issues. Even when I try to totally uninstall and re-install, it still is not working for me. One thing I'm noticing is that it doesn't have a Schema version in the project config file:

```
snaptcha:
  enabled: true
  licenseKey: XXX
  settings:
    allowList: ''
    denyList: ''
    errorButtonText: Resubmit
    errorJavascriptDisabled: 'JavaScript must be enabled for your submission to be verified. Please enable it and refresh the page.'
    errorMessage: 'For some reason your submission was flagged as spam. Please click the button below to resubmit your data.'
    errorTemplate: ''
    errorTitle: Error
    excludeControllerActions: ''
    expirationTime: '60'
    fieldName: snaptcha
    logRejected: '1'
    minimumSubmitTime: '1'
    oneTimeKey: ''
    salt: XXX
    validateUsers: '1'
    validationEnabled: '1'
```

### To reproduce
Steps to reproduce the behaviour:
1. Install snaptcha.
2. Try and use a form that has a snaptcha tag.
3. Get an error.

### Versions
- Plugin version: 3.0.2
- Craft version: 3.6.12

**Comment by codyjames** [comment]
I was able to resolve this by manually deleting the tables on my staging and production environments. Migrations not firing is kinda a huge pain.

**Comment by bencroker** [comment]
> Even when I try to totally uninstall and re-install, it still is not working for me.

That sounds very unusual. Glad to hear you managed to make it work in the end!

**Comment by stenvdb** [comment]
I had the same issue. Had to manually uninstall and re-install the plugin on live environment. Same issue: migrations that didn't run. Sounds like https://github.com/putyourlightson/craft-snaptcha/issues/14 has the same issue. Guessing there is something wrong as it does not seems to be very unusual when 3 people have the same issue?

**Comment by bencroker** [comment]
I agree. If you could open a new issue with details of what version you updated from and the steps to reproduce then we can look into fixing it.

**Comment by codyjames** [comment]
I could be totally wrong here but I wonder if it has to do with schemaVersion and this line: https://github.com/putyourlightson/craft-snaptcha/blob/728e3709d9bf72f8b11f5631c5c0683d85fd43ab/src/migrations/m210217_120000_add_value_column.php#L23

For whatever reason, I didn't have a schemaVersion value in my projectConfig for Snaptcha. Could this have caused the migration to not run?

**Comment by bencroker** [comment]
Possibly @codyjames, but you haven't shown me the error message you received, so it's almost impossible to know what the issue was.

**Comment by codyjames** [comment]
```
yii\base\UnknownPropertyException: Setting unknown property: putyourlightson\snaptcha\records\SnaptchaRecord::value
#34 /var/www/html/vendor/yiisoft/yii2/base/Component.php(210): yii\base\Component::__set
#33 /var/www/html/vendor/yiisoft/yii2/db/BaseActiveRecord.php(324): yii\db\BaseActiveRecord::__set
#32 /var/www/html/vendor/craftcms/cms/src/db/ActiveRecord.php(54): craft\db\ActiveRecord::__set
#31 /var/www/html/vendor/yiisoft/yii2/BaseYii.php(558): yii\BaseYii::configure
#30 /var/www/html/vendor/yiisoft/yii2/base/BaseObject.php(107): yii\base\BaseObject::__construct
#29 /var/www/html/vendor/putyourlightson/craft-snaptcha/src/services/SnaptchaService.php(294): putyourlightson\snaptcha\services\SnaptchaService::_getSnaptchaRecord
#28 /var/www/html/vendor/putyourlightson/craft-snaptcha/src/services/SnaptchaService.php(52): putyourlightson\snaptcha\services\SnaptchaService::getFieldKey
#27 /var/www/html/vendor/putyourlightson/craft-snaptcha/src/variables/SnaptchaVariable.php(54): putyourlightson\snaptcha\variables\SnaptchaVariable::getField
#26 /var/www/html/vendor/twig/twig/src/Extension/CoreExtension.php(1507): twig_get_attribute
#25 /var/www/html/vendor/craftcms/cms/src/helpers/Template.php(106): craft\helpers\Template::attribute
#24 /var/www/html/storage/runtime/compiled_templates/9a/9adfcbb6bdf3dfb36f8a9f6ce6f0a726e7045911263bf2dadd767dedf6f41ac9.php(103): __TwigTemplate_b172ed5d39327e53c5b85ca3f5b9b1b0eef9c66f703d7bd3fcf42b5f89a9654c::block_content
#23 /var/www/html/vendor/twig/twig/src/Template.php(182): Twig\Template::displayBlock
#22 /var/www/html/storage/runtime/compiled_templates/b5/b5d4930783b9cb7a32f3f1f7ed744d67a12f7f9a59983bb11ce553d4b95991a8.php(207): __TwigTemplate_b25931a9607e675dc8f51859b42ad6f9cc012ae454f2cf52095f28d4101fff10::doDisplay
#21 /var/www/html/vendor/twig/twig/src/Template.php(405): Twig\Template::displayWithErrorHandling
#20 /var/www/html/vendor/twig/twig/src/Template.php(378): Twig\Template::display
#19 /var/www/html/storage/runtime/compiled_templates/9a/9adfcbb6bdf3dfb36f8a9f6ce6f0a726e7045911263bf2dadd767dedf6f41ac9.php(51): __TwigTemplate_b172ed5d39327e53c5b85ca3f5b9b1b0eef9c66f703d7bd3fcf42b5f89a9654c::doDisplay
#18 /var/www/html/vendor/twig/twig/src/Template.php(405): Twig\Template::displayWithErrorHandling
#17 /var/www/html/vendor/twig/twig/src/Template.php(378): Twig\Template::display
#16 /var/www/html/vendor/twig/twig/src/Template.php(390): Twig\Template::render
#15 /var/www/html/vendor/twig/twig/src/TemplateWrapper.php(45): Twig\TemplateWrapper::render
#14 /var/www/html/vendor/twig/twig/src/Environment.php(318): Twig\Environment::render
#13 /var/www/html/vendor/craftcms/cms/src/web/View.php(390): craft\web\View::renderTemplate
#12 /var/www/html/vendor/craftcms/cms/src/web/View.php(451): craft\web\View::renderPageTemplate
#11 /var/www/html/vendor/craftcms/cms/src/web/Controller.php(257): craft\web\Controller::renderTemplate
#10 /var/www/html/vendor/craftcms/cms/src/controllers/TemplatesController.php(100): craft\controllers\TemplatesController::actionRender
#9 [internal](0): call_user_func_array
#8 /var/www/html/vendor/yiisoft/yii2/base/InlineAction.php(57): yii\base\InlineAction::runWithParams
#7 /var/www/html/vendor/yiisoft/yii2/base/Controller.php(181): yii\base\Controller::runAction
#6 /var/www/html/vendor/craftcms/cms/src/web/Controller.php(190): craft\web\Controller::runAction
#5 /var/www/html/vendor/yiisoft/yii2/base/Module.php(534): yii\base\Module::runAction
#4 /var/www/html/vendor/craftcms/cms/src/web/Application.php(274): craft\web\Application::runAction
#3 /var/www/html/vendor/yiisoft/yii2/web/Application.php(104): yii\web\Application::handleRequest
#2 /var/www/html/vendor/craftcms/cms/src/web/Application.php(259): craft\web\Application::handleRequest
#1 /var/www/html/vendor/yiisoft/yii2/base/Application.php(392): yii\base\Application::run
#0 /index.php(21): null
```

**Comment by bencroker** [comment]
Yup, that does indeed seem to indicate that the `m210217_120000_add_value_column` migration didn't run. Will look into it, thanks!

**Comment by bencroker** [comment]
I'm still unable to replicated this, even after manually deleting the `schemaVersion` from project config and following the steps you describe. I'm going to close this for now and if it is reported again then will have to dig deeper.

**Comment by tobystokes** [comment]
fwiw, had the same trouble upgrading from
"2.3.1", to "3.0.2",

specifically, seemed to work fine on local dev, but migration didn't run in live env.
errors like
`info 2021-05-06 14:11:20 [-][-][b8f1661d6b9c8b549e900273344e207c][error][yii\base\UnknownPropertyException] yii\base\UnknownPropertyException: Setting unknown property: putyourlightson\snaptcha\records\SnaptchaRecord::value in /var/www/html/vendor/yiisoft/yii2/base/Component.php:210 `

**Comment by bencroker** [comment]
@tobystokes Can you check whether a `schemaVersion` value for Snaptcha exists in your `project.yaml` file?

**Comment by tobystokes** [comment]
> @tobystokes Can you check whether a `schemaVersion` value for Snaptcha exists in your `project.yaml` file?

The marvels of git! 
it was updated from     
`2.1.0` to `3.0.0` on my first local `./craft update all`

I then deployed to live env at https://servd.host/ (which I think basically does a `craft migrate/all` & `craft project-config/apply`) but no effect.

I also tried a complete plugin remove and reinstall - I noticed after fresh install there was NO `schemaVersion` `project.yaml`. Redeploying this didn't seem to run the migration either.

I have since just recreated the table on live, which was no biggy seeing as it was empty anyway, and that's got it working again.

**Comment by bencroker** [comment]
Thanks, added a fix in https://github.com/putyourlightson/craft-snaptcha/commit/9bfacfc5767750b01de22d0dc1daa59f792928b0 for the next release.

**Comment by bencroker** [comment]
Released in version 3.0.5 which should hopefully resolve this.

**Comment by sebastian-lenz** [comment]
I just ran into the very same issue when upgrading from 2.3.1 to 3.0.5. The update ran well on my development system but when deploying the update to the production system, the table does not get updated. The migration `m210217_120000_add_value_column` shows up in the table `migrations` but the table `snaptcha` is unchanged and the column `value` is missing.

I guess it is not a good idea to rely on the `schemaVersion` value as it is stored within the config files which are not guaranteed the be in sync with the database.

**Comment by leobauza** [comment]
Ran into this issue as well updating from `2.2.5` to `3.0.5`. Same problem @sebastian-lenz described. Full error below:

```
yii\base\UnknownPropertyException: Setting unknown property: putyourlightson\snaptcha\records\SnaptchaRecord::value in /storage/av06163/www/public_html/releases/20210603122534/vendor/yiisoft/yii2/base/Component.php:210
Stack trace:
#0 /storage/av06163/www/public_html/releases/20210603122534/vendor/yiisoft/yii2/db/BaseActiveRecord.php(324): yii\base\Component->__set()
#1 /storage/av06163/www/public_html/releases/20210603122534/vendor/craftcms/cms/src/db/ActiveRecord.php(54): yii\db\BaseActiveRecord->__set()
#2 /storage/av06163/www/public_html/releases/20210603122534/vendor/yiisoft/yii2/BaseYii.php(558): craft\db\ActiveRecord->__set()
#3 /storage/av06163/www/public_html/releases/20210603122534/vendor/yiisoft/yii2/base/BaseObject.php(107): yii\BaseYii::configure()
#4 /storage/av06163/www/public_html/releases/20210603122534/vendor/putyourlightson/craft-snaptcha/src/services/SnaptchaService.php(294): yii\base\BaseObject->__construct()
#5 /storage/av06163/www/public_html/releases/20210603122534/vendor/putyourlightson/craft-snaptcha/src/services/SnaptchaService.php(52): putyourlightson\snaptcha\services\SnaptchaService->_getSnaptchaRecord()
#6 /storage/av06163/www/public_html/releases/20210603122534/vendor/putyourlightson/craft-snaptcha/src/variables/SnaptchaVariable.php(54): putyourlightson\snaptcha\services\SnaptchaService->getFieldKey()
#7 /storage/av06163/www/public_html/releases/20210603122534/vendor/twig/twig/src/Extension/CoreExtension.php(1511): putyourlightson\snaptcha\variables\SnaptchaVariable->getField()
#8 /storage/av06163/www/public_html/releases/20210603122534/vendor/craftcms/cms/src/helpers/Template.php(106): twig_get_attribute()
#9 /storage/av06163/www/public_html/shared/storage/runtime/compiled_templates/f4/f4e9aab8cdb9deff600ba2240281ce24574a8aab24063e5695128b26652cf276.php(113): craft\helpers\Template::attribute()
#10 /storage/av06163/www/public_html/releases/20210603122534/vendor/twig/twig/src/Template.php(405): __TwigTemplate_337a90865868627e53194ed0b3093cc87acdc282851ce13d1d3a9b3e9a102c7f->doDisplay()
#11 /storage/av06163/www/public_html/releases/20210603122534/vendor/twig/twig/src/Template.php(378): Twig\Template->displayWithErrorHandling()
#12 /storage/av06163/www/public_html/releases/20210603122534/vendor/twig/twig/src/Template.php(390): Twig\Template->display()
#13 /storage/av06163/www/public_html/releases/20210603122534/vendor/twig/twig/src/TemplateWrapper.php(45): Twig\Template->render()
#14 /storage/av06163/www/public_html/releases/20210603122534/vendor/twig/twig/src/Environment.php(318): Twig\TemplateWrapper->render()
#15 /storage/av06163/www/public_html/releases/20210603122534/vendor/craftcms/cms/src/web/View.php(389): Twig\Environment->render()
#16 /storage/av06163/www/public_html/releases/20210603122534/vendor/viget/craft-viget-base/src/services/PartialLoader.php(32): craft\web\View->renderTemplate()
#17 /storage/av06163/www/public_html/shared/storage/runtime/compiled_templates/f2/f250b8a7fd8400b5c42260c9c496e7e19614d5d0a55d881f279f6dc77ccc883f.php(124): viget\base\services\PartialLoader::load()
#18 /storage/av06163/www/public_html/releases/20210603122534/vendor/twig/twig/src/Template.php(182): __TwigTemplate_bd3e6e618f97499ad3b3f7fe3bd1746cd64b378fe528a7a353c1af2018161fb9->block_content()
#19 /storage/av06163/www/public_html/shared/storage/runtime/compiled_templates/94/94be84ea0566e9d0a396fdf30e1d0c5dac23e3731a3312a542a48955eea130c6.php(313): Twig\Template->displayBlock()
#20 /storage/av06163/www/public_html/releases/20210603122534/vendor/twig/twig/src/Template.php(405): __TwigTemplate_6c127337604a21d105fbaedc413d1dfc1784b0ade2fbc5a85a23799147b7a0fa->doDisplay()
#21 /storage/av06163/www/public_html/releases/20210603122534/vendor/twig/twig/src/Template.php(378): Twig\Template->displayWithErrorHandling()
#22 /storage/av06163/www/public_html/shared/storage/runtime/compiled_templates/f2/f250b8a7fd8400b5c42260c9c496e7e19614d5d0a55d881f279f6dc77ccc883f.php(50): Twig\Template->display()
#23 /storage/av06163/www/public_html/releases/20210603122534/vendor/twig/twig/src/Template.php(405): __TwigTemplate_bd3e6e618f97499ad3b3f7fe3bd1746cd64b378fe528a7a353c1af2018161fb9->doDisplay()
#24 /storage/av06163/www/public_html/releases/20210603122534/vendor/twig/twig/src/Template.php(378): Twig\Template->displayWithErrorHandling()
#25 /storage/av06163/www/public_html/releases/20210603122534/vendor/twig/twig/src/Template.php(390): Twig\Template->display()
#26 /storage/av06163/www/public_html/releases/20210603122534/vendor/twig/twig/src/TemplateWrapper.php(45): Twig\Template->render()
#27 /storage/av06163/www/public_html/releases/20210603122534/vendor/twig/twig/src/Environment.php(318): Twig\TemplateWrapper->render()
#28 /storage/av06163/www/public_html/releases/20210603122534/vendor/craftcms/cms/src/web/View.php(389): Twig\Environment->render()
#29 /storage/av06163/www/public_html/releases/20210603122534/vendor/craftcms/cms/src/web/View.php(450): craft\web\View->renderTemplate()
#30 /storage/av06163/www/public_html/releases/20210603122534/vendor/craftcms/cms/src/web/Controller.php(257): craft\web\View->renderPageTemplate()
#31 /storage/av06163/www/public_html/releases/20210603122534/vendor/craftcms/cms/src/controllers/TemplatesController.php(100): craft\web\Controller->renderTemplate()
#32 [internal function]: craft\controllers\TemplatesController->actionRender()
#33 /storage/av06163/www/public_html/releases/20210603122534/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array()
#34 /storage/av06163/www/public_html/releases/20210603122534/vendor/yiisoft/yii2/base/Controller.php(181): yii\base\InlineAction->runWithParams()
#35 /storage/av06163/www/public_html/releases/20210603122534/vendor/craftcms/cms/src/web/Controller.php(190): yii\base\Controller->runAction()
#36 /storage/av06163/www/public_html/releases/20210603122534/vendor/yiisoft/yii2/base/Module.php(534): craft\web\Controller->runAction()
#37 /storage/av06163/www/public_html/releases/20210603122534/vendor/craftcms/cms/src/web/Application.php(278): yii\base\Module->runAction()
#38 /storage/av06163/www/public_html/releases/20210603122534/vendor/yiisoft/yii2/web/Application.php(104): craft\web\Application->runAction()
#39 /storage/av06163/www/public_html/releases/20210603122534/vendor/craftcms/cms/src/web/Application.php(263): yii\web\Application->handleRequest()
#40 /storage/av06163/www/public_html/releases/20210603122534/vendor/yiisoft/yii2/base/Application.php(392): craft\web\Application->handleRequest()
#41 /storage/av06163/www/public_html/releases/20210603122534/public/index.php(21): yii\base\Application->run()
#42 {main}
```

**Comment by bencroker** [comment]
Just released 3.0.6 which should resolve this issue once and for all.
